### DISCUSSION TITLE
Feature that checks the quality of code already written

### DISCUSSION THREAD

**Discussion by ollie-iterators** [community/community#7110] [Copilot Conversations]
An interesting feature that could be added to GitHub Copilot is something that checks out the code that has already been written. Some examples of things that could be checked are:

1. The naming of functions or variables
2. The writing of comments
3.  How code is written (This might need to happen when it is able to detect bugs (maybe through codeQL), or code quality (problably through an automatic format checker added to GitHub Copilot), because I have read that sometimes the code written is unsafe and may not work)

**Comment by avigloz** [comment]
Why not just use a linter (or similar plugin)? As far as I understand, it's pretty difficult to train an AI (and even humans) to determine "quality" objectively, especially in a space as ambiguous as code (where quality standards vary heavily from codebase to codebase and per organization, not to mention per language, etc.)

**Reply by avigloz** [reply]
Namely, something like SonarLint (https://www.sonarlint.org/, free and OSS) is probably sufficient.

**Reply by ollie-iterators** [reply]
Ok, so the feature should be making GitHub Copilot be able to combine it's suggestions with a linter (or similar plugin) and have some suggestions for good linters/plugins for the various programming languages that work with GitHub Copilot.

I just think that it would be a good way to help the people who write the code be able to make the code better so that Copilot could improve in a virtuous cycle.

**Reply by ollie-iterators** [reply]
Also, some way to work on naming variables and functions could be encouraging people to add comments describing what the function or variable is for and seeing if the name of the variable matches what it is said to do.

**Reply by avigloz** [reply]
> could be encouraging people to add comments describing what the function or variable is for and seeing if the name of the variable matches what it is said to do

As I mentioned, standards vary across codebases, and have to do with function sizes, comments, naming, cyclomatic complexity, etc. Some standards (namely the prolific "Clean Code" set of methodologies) outright discourage commenting and have other strong stances, and for good reason. I think it would be frustrating to developers to have some arbitrary set of standards imposed on them via Copilot's suggestions.

IMHO Copilot should not *intentionally* encourage the application of any particular set of coding standards (or, it should keep doing what its doing and building its suggestions from public code - if standards somehow change uniformly across the entire industry, then that would drive change on its own). Otherwise, things like comments, naming, etc. (and how exactly these things are done) should be solely dependent on an individual developer/org's choices and diligence.

**Reply by ollie-iterators** [reply]
Ok, if variables and function names recommendation is too far, then that recommendation should not be done. I still think that there are probably some good ways to try to nudge people into improving the code that they write, which could be by encouraging them to get rid of duplicate code. After all, researchers have found that more than half of GitHub is duplicate code. https://www.theregister.com/2017/11/21/github_duplicate_code/#:~:text=Given%20that%20code%20sharing%20is,cent%2C%20a%20study%20has%20found.

**Reply by avigloz** [reply]
Sure, that sounds better! It would be a good start for Copilot to not regurgitate code it found on the internet, and instead tailors it more closely to the codebase being used (similar to how Tabnine learns your coding patterns/activities, then essentially saves you keystrokes while also suggesting novel code based on the way you code).

**Reply by ollie-iterators** [reply]
GitHub Copilot also already works like Clean Code, because if you make the variables and functions describe what you want them to do, it says that it works better.
